Added a new doc for adding projects to the repo

Daniel-Genkin-MS-2 opened the pull request:
I recently added a new project for Middleware related to some other changed to the repo. I noticed though that there isn't any documentation on how to do this. As we can't always ask Will for help :), this PR adds a new file which explains how to create new projects in the repo.

Please let me know if there is anything that I missed or that should be added.

Daniel-Genkin-MS-2 committed:
Added a new doc for addign projects to the repo

captainsafia commented on docs/AddingNewProjects.md:
```diff
@@ -0,0 +1,44 @@
+# Adding New Projects To The Repo
```
Can you add an entry for this new document to the table of contents in `docs/README.dm`?

wtgodbe commented on docs/AddingNewProjects.md:
```diff
@@ -0,0 +1,44 @@
+  - This only lists APIs that have NOT already been shipped to customers. So, these can still change.
+
+## Adding to the rest of the repo
+- VS Should have already registered your CS projec in the corresponding solution (`.sln`) and solution filter (`.slnf`) files. See this [Example](https://github.com/dotnet/aspnetcore/pull/41945/files#diff-cd977e0a76b37d35c04d9d819ea66ef8a35d9ef7f86a9a7c774d751e8119db4fR1713-R11118)
```
CS projec -> .csproj

wtgodbe commented on docs/AddingNewProjects.md:
````diff
@@ -0,0 +1,44 @@
+    ```XML
+    <ProjectReferenceProvider Include="[Your project id]" ProjectPath="$(RepoRoot)src\[Rest of the path to your project]" />
+    ```
+- Add the following new line to the `eng/SharedFramework.Local.props` file to include your new project:
````
This is only needed if you set `IsAspNetCoreApp` to true

wtgodbe reviewed (changes requested):
Looks mostly good other than the few comments - thanks for this!

captainsafia commented on docs/AddingNewProjects.md:
```diff
@@ -0,0 +1,44 @@
+1. Create a new folder that will house your `.csproj` and other project related files.
+2. In VS, Add a new `Solution Folder` in the same place as the new folder from the previous step.
+3. Create the project via the VS `Add` menu (select the folder -> right click -> Add -> follow the wizard).
+4. Add the following `ItemGroup` to your newly generated `.csproj`:
```
Nit: A brief explainer to outline that this is for interacting with the PublicAPI analyzer?

Daniel-Genkin-MS-2 commented on docs/AddingNewProjects.md:
```diff
@@ -0,0 +1,44 @@
+1. Create a new folder that will house your `.csproj` and other project related files.
+2. In VS, Add a new `Solution Folder` in the same place as the new folder from the previous step.
+3. Create the project via the VS `Add` menu (select the folder -> right click -> Add -> follow the wizard).
+4. Add the following `ItemGroup` to your newly generated `.csproj`:
```
Are we going for minimal descriptions? I was trying to cover everything as removing is easier than adding. Do you want to remove the whole section or just the obvious parts?

Daniel-Genkin-MS-2 replied:
Removed

captainsafia commented on docs/AddingNewProjects.md:
```diff
@@ -0,0 +1,44 @@
+  - This only lists APIs that have NOT already been shipped to customers. So, these can still change.
+
+## Adding to the rest of the repo
+- VS Should have already registered your CS projec in the corresponding solution (`.sln`) and solution filter (`.slnf`) files. See this [Example](https://github.com/dotnet/aspnetcore/pull/41945/files#diff-cd977e0a76b37d35c04d9d819ea66ef8a35d9ef7f86a9a7c774d751e8119db4fR1713-R11118)
```
Is there a way to link to an existing commit in the repo instead of a diff in a PR?

Daniel-Genkin-MS-2 replied:
Is this better: https://github.com/dotnet/aspnetcore/pull/41945/commits/d1fddbfbac25c529e153a08ca1a694f32a69482e

Daniel-Genkin-MS-2 replied:
```
https://github.com/dotnet/aspnetcore/pull/41945/commits/d1fddbfbac25c529e153a08ca1a694f32a69482e
```
^The actual link without Github magic

Daniel-Genkin-MS-2 committed:
Added to outline + some fixes

Daniel-Genkin-MS-2 committed:
Clarified about public api analyzers

Daniel-Genkin-MS-2 committed:
Added a new step that I just discovered is necessary

Daniel-Genkin-MS-2 committed:
Upon discovery of a script that does half the work for you, I have simplified this document

Daniel-Genkin-MS-2 committed:
Removed unecessary step

Daniel-Genkin-MS-2 committed:
updated example link to be a commit rather than a diff

wtgodbe commented on docs/AddingNewProjects.md:
```diff
@@ -0,0 +1,31 @@
+  - This only lists APIs that have NOT already been shipped to customers. So, these can still change.
+
+## Adding to the rest of the repo
+1. VS Should have already registered your `.csproj` in the corresponding solution (`.sln`) and solution filter (`.slnf`) files. See this [Example](https://github.com/dotnet/aspnetcore/pull/41945/commits/d1fddbfbac25c529e153a08ca1a694f32a69482e)
```
This still links to a commit where the .csproj is not inside of a `src` folder

wtgodbe commented on docs/AddingNewProjects.md:
```diff
@@ -0,0 +1,31 @@
+## Creating a new project
+1. Create a new folder that will house your `.csproj` and other project-related files.
+2. (EXTREMELY IMPORTANT) Inside our new folder make a new folder called `src`.
+3. In VS, Add a new `Solution Folder` in the new `src` folder from the previous step.
```
I think you want to add the solution folder from the folder you created in step 1, so that it will also include `tests` or `samples` you add there - can someone else confirm (@BrennanConroy @halter73 @Tratcher)? I realized recently I've been doing this wrong, I had manually been adding each .csproj (src, test, sample) one-by-one.

wtgodbe replied:
I mean that this step says to add the solution folder in `src/Middleware/WebTransport/src/` when it should actually be in `src/Middleware/WebTransport/`

Daniel-Genkin-MS-2 replied:
Oh, ok. Just pushed the change. Thanks

wtgodbe commented on docs/AddingNewProjects.md:
```diff
@@ -0,0 +1,31 @@
+Sample PR of final result: https://github.com/dotnet/aspnetcore/pull/41945
+
+## Creating a new project
+1. Create a new folder that will house your `.csproj` and other project-related files.
```
Worth noting here that if this is a new folder in https://github.com/dotnet/aspnetcore/tree/main/src, you need to add a reference to it to eng\build.props:

https://github.com/dotnet/aspnetcore/blob/41e03fe5995cd59213056179343df168228650ce/eng/Build.props#L160-L168

And

https://github.com/dotnet/aspnetcore/blob/41e03fe5995cd59213056179343df168228650ce/eng/Build.props#L211-L219

wtgodbe replied:
Otherwise build.cmd won't know to build it. Your new project is inside of `src\Middleware`, so it's already included automatically & you didn't need to do this step.

Daniel-Genkin-MS-2 committed:
Updated the link

Daniel-Genkin-MS-2 committed:
Added explainer about the build.props file

Daniel-Genkin-MS-2 committed:
Adjustment to step 3 of creating a new project

wtgodbe reviewed (approved):
LGTM other than the submodule conflict

Daniel-Genkin-MS-2 committed:
Updated submodule

Daniel-Genkin-MS-2 committed:
Merge branch 'main' of https://github.com/dotnet/aspnetcore into t-dagenkin/Adding-A-CSProj-doc

Daniel-Genkin-MS-2 commented:
@captainsafia @dougbu Do you have any comments or suggestions before I turn on auto-merge?

dougbu commented:
> Do you have any more comments or suggestions? I plan to turn on auto merge soon.

Could you wait 'til tonight❔ I'm on build ops and would like to double check this after working hours. I can enable auto-merge once I'm done.

Daniel-Genkin-MS-2 commented:
> > Do you have any more comments or suggestions? I plan to turn on auto merge soon.
> 
> Could you wait 'til tonight❔ I'm on build ops and would like to double check this after working hours. I can enable auto-merge once I'm done.

Yes, sorry I didn't mean to rush anyone. I just wanted to verify if you had any more comments. So, this plan sounds good to me. Also feel free to do it tomorrow if that would be easier for you.

Daniel-Genkin-MS-2 committed:
Just learned about a test that is probably worth mentioning here as it breaks when you add to sharedFx

Daniel-Genkin-MS-2 commented:
Sorry everyone for pushing new code/content after some have already approved. I just discovered a CI test that fails when you add to sharedFx. So, I just pushed a commit to include a message about it.

dougbu reviewed (commented):
Lots of comments but mostly nit-picking. My main overall comment is that it doesn't really cover adding non-`src/*.csproj` projects e.g. new test/ projects in an existing folder under our top-level src/. Lots of the steps are useless in that case.

Daniel-Genkin-MS-2 commented on docs/AddingNewProjects.md:
```diff
@@ -0,0 +1,39 @@
+
+## Adding to the rest of the repo
+1. VS Should have already registered your `.csproj` in the corresponding solution (`.sln`) and solution filter (`.slnf`) files. See this [Example](https://github.com/dotnet/aspnetcore/pull/41945/commits/586ccc8c895862b65645c4b0f979db1eecd29626)
+  - If VS has not already modified these files, make sure to add it manually as is visible in the example listed above.
```
Is this possible? Since VS generates all the unique ids for everything, it would be quite difficult to make sure that the ids are unique and match up. Having said this, this is why I included the links.

Daniel-Genkin-MS-2 committed:
Thank you Doug for your suggestions

Daniel-Genkin-MS-2 commented:
Wow, thank you for pointing these things out. I completely missed them. I also agree that moving all the SharedFx stuff into a designated section makes the doc easier to read.

dougbu commented on docs/AddingNewProjects.md:
```diff
@@ -0,0 +1,48 @@
+    - `PublicAPI.Shipped.txt`
+      - Lists publically visible APIs that are exported from your final compiled `.dll`.
+      - This only lists APIs that have already been shipped to customers and cannot be changed.
+      - There is an empty template at `eng/PublicAPI.empty.txt` for your reference.
```
Instead, recommend copying the file into the right location and renaming it.

dougbu commented on docs/AddingNewProjects.md:
```diff
@@ -0,0 +1,48 @@
+      - This only lists APIs that have already been shipped to customers and cannot be changed.
+      - There is an empty template at `eng/PublicAPI.empty.txt` for your reference.
+    - `PublicAPI.UnShipped.txt`
+      - Lists publicly visible APIs that are exported from your final compiled `.dll`. If this is not configured properly, you will get build errors. VS will warn you though with green squiggly lines. If you see these squiggly lines, open the VS Quick Actions (CTRL + '.') and select the option to and it to the public API.
```
👍

dougbu commented on docs/AddingNewProjects.md:
```diff
@@ -0,0 +1,48 @@
+    - `PublicAPI.UnShipped.txt`
+      - Lists publicly visible APIs that are exported from your final compiled `.dll`. If this is not configured properly, you will get build errors. VS will warn you though with green squiggly lines. If you see these squiggly lines, open the VS Quick Actions (CTRL + '.') and select the option to and it to the public API.
+      - This only lists APIs that have NOT already been shipped to customers. So, these can still change.
+      - There is an empty template at `eng/PublicAPI.empty.txt` for your reference.
```
Again, copying is easier.

dougbu commented on docs/AddingNewProjects.md:
```diff
@@ -0,0 +1,48 @@
+      - There is an empty template at `eng/PublicAPI.empty.txt` for your reference.
+    - `AssemblyInfo.cs`
+      - Lists various properties of your compiled assembly, such as to which other packages your `internal` properties and methods are available to.
+      - You can also expose internals via the `@(InternalsVisibleTo)` item in your project file instead of using this `AssemblyInfo.cs` file.
```
nit
```suggestion
      - You can also expose internals via `@(InternalsVisibleTo)` items in your project file instead of using this `AssemblyInfo.cs` file.
```

dougbu reviewed (commented):
Reads much better to me. Thanks @Daniel-Genkin-MS-2

Daniel-Genkin-MS-2 committed:
Addressed more PR comments

captainsafia commented on docs/AddingNewProjects.md:
```diff
@@ -0,0 +1,51 @@
+      - You can also expose internals via `@(InternalsVisibleTo)` items in your project file instead of using this `AssemblyInfo.cs` file.
+
+## Adding to the rest of the repo
+1. VS Should have already registered your `.csproj` in the corresponding solution ([`.sln`](https://github.com/dotnet/aspnetcore/blob/586ccc8c895862b65645c4b0f979db1eecd29626/AspNetCore.sln)) and solution filter ([`.slnf`](https://github.com/dotnet/aspnetcore/blob/586ccc8c895862b65645c4b0f979db1eecd29626/src/Middleware/Middleware.slnf#L107-L109)) files.
```
```suggestion
1. VS should have already registered your `.csproj` in the corresponding solution ([`.sln`](https://github.com/dotnet/aspnetcore/blob/586ccc8c895862b65645c4b0f979db1eecd29626/AspNetCore.sln)) and solution filter ([`.slnf`](https://github.com/dotnet/aspnetcore/blob/586ccc8c895862b65645c4b0f979db1eecd29626/src/Middleware/Middleware.slnf#L107-L109)) files.
```

captainsafia commented on docs/AddingNewProjects.md:
```diff
@@ -0,0 +1,51 @@
+      - There is an empty template at `eng/PublicAPI.empty.txt` for your reference. You can copy and rename the file to add it to your project. Make sure the name is exactly as shown above.
+    - `AssemblyInfo.cs`
+      - Lists various properties of your compiled assembly, such as to which other packages your `internal` properties and methods are available to.
+      - You can also expose internals via `@(InternalsVisibleTo)` items in your project file instead of using this `AssemblyInfo.cs` file.
```
In the past, we've talked about being consistent about using the `InternalsVisibleTo` property vs. AssemblyInfo for internals. I think we should lean towards recommending people to use the `@(InternalsVisibleTo)` here so we can be more consistent in the future.

Daniel-Genkin-MS-2 replied:
Ok. How do you recommend I update this? Should I just remove this bullet about `AssemblyInfo.cs` and add your comment as a note under the `Adding to the rest of the repo` section? I just think that this point seems to deviate a bit from the main objective of this particular doc. Is there a better place to put this that might be more directly applicable to the contents of csproj files or the `AssemblyInfo.cs` file?

Daniel-Genkin-MS-2 committed:
Update docs/AddingNewProjects.md

Co-authored-by: Safia Abdalla <safia@safia.rocks>

dougbu commented on docs/AddingNewProjects.md:
```diff
@@ -0,0 +1,51 @@
+
+## Creating a new project
+1. Create a new folder that will house your `.csproj` and other project-related files.
+2. (EXTREMELY IMPORTANT) Inside our new folder make a new folder called `src`.
```
I still suggest making it clear `src/` is just one option. When adding a `test/` project or similar the issues in (4) do not apply.

dougbu commented on docs/AddingNewProjects.md:
````diff
@@ -0,0 +1,51 @@
+  ```
+
+## Including your project in SharedFx
+1. (OPTIONAL) Add the following line to the `.csproj`'s `PropertyGroup` to include your project in the SharedFx API:
````
Minor but this section is entirely optional, not just the first line

dougbu reviewed (commented):
This document is starting to look really solid❕

Daniel-Genkin-MS-2 committed:
Addressed more PR comments

Daniel-Genkin-MS-2 committed:
Merge branch 't-dagenkin/Adding-A-CSProj-doc' of https://github.com/Daniel-Genkin-MS-2/aspnetcore into t-dagenkin/Adding-A-CSProj-doc

Daniel-Genkin-MS-2 commented:
@dougbu I agree. Thank you for your feedback everyone! Any more suggestions or may I merge?

halter73 commented on docs/AddingNewProjects.md:
```diff
@@ -0,0 +1,51 @@
+      - Lists publicly visible APIs that are exported from your final compiled `.dll`. If this is not configured properly, you will get build errors. VS will warn you though with green squiggly lines. If you see these squiggly lines, open the VS Quick Actions (CTRL + '.') and select the option to and it to the public API.
+      - This only lists APIs that have NOT already been shipped to customers. So, these can still change.
+      - There is an empty template at `eng/PublicAPI.empty.txt` for your reference. You can copy and rename the file to add it to your project. Make sure the name is exactly as shown above.
+    - `AssemblyInfo.cs`
+      - Lists various properties of your compiled assembly, such as to which other packages your `internal` properties and methods are available to.
+      - You can also expose internals via `@(InternalsVisibleTo)` items in your project file instead of using this `AssemblyInfo.cs` file.
```
```suggestion
    - You can expose internals via `@(InternalsVisibleTo)` items in your `.csproj`.
      ```XML
      <ItemGroup>
        <InternalsVisibleTo Include="Microsoft.AspNetCore.My.TestProject" />
      ```
```

To @captainsafia's point, we shouldn't need to mention `AssemblyInfo.cs` at all since we no longer recommend it for `InternalsVisibleTo`.

Also, are we using 2 or 4 spaces for indentation? We seem to be inconsistent with this.

halter73 commented on docs/AddingNewProjects.md:
```diff
@@ -0,0 +1,51 @@
+      - You can also expose internals via `@(InternalsVisibleTo)` items in your project file instead of using this `AssemblyInfo.cs` file.
+
+## Adding to the rest of the repo
+1. VS should have already registered your `.csproj` in the corresponding solution ([`.sln`](https://github.com/dotnet/aspnetcore/blob/586ccc8c895862b65645c4b0f979db1eecd29626/AspNetCore.sln)) and solution filter ([`.slnf`](https://github.com/dotnet/aspnetcore/blob/586ccc8c895862b65645c4b0f979db1eecd29626/src/Middleware/Middleware.slnf#L107-L109)) files.
```
This is the first time the `.sln` or .`slnf` was mentioned. Why should they already be updated? Unless you opened an existing `.sln` or `.slnf` in the first place, I don't think it would be unless we're talking about a *new* `.sln` which we don't want.

The very start of this document should make it clear what we're opening in VS. If someone just does "Create a new project" outside the context of an existing solution, VS will helpfully (or unhelpfully in our case) try to crate a new `.sln` for you.

halter73 replied:
I think the edit I suggested in https://github.com/dotnet/aspnetcore/pull/41949#discussion_r891641435 is probably sufficient for this.

halter73 commented on docs/AddingNewProjects.md:
```diff
@@ -0,0 +1,51 @@
+
+## Adding to the rest of the repo
+1. VS should have already registered your `.csproj` in the corresponding solution ([`.sln`](https://github.com/dotnet/aspnetcore/blob/586ccc8c895862b65645c4b0f979db1eecd29626/AspNetCore.sln)) and solution filter ([`.slnf`](https://github.com/dotnet/aspnetcore/blob/586ccc8c895862b65645c4b0f979db1eecd29626/src/Middleware/Middleware.slnf#L107-L109)) files.
+  - If VS has not already modified these files, make sure to add it manually as is visible in the example listed above.
```
If you did create a project outside of `AspNetCore.sln` adding it manually with a plain text editor is nearly impossible. This should mention that adding an "Existing Project..." in VS

```suggestion
  - If VS has not already modified these files, open the `.slnf` you want to add the project to. Create a solution folder for your project it if doesn't exist already. Then right click solution foulder -> Add -> Existing Project... -> follow the wizard.
```

halter73 commented on docs/AddingNewProjects.md:
```diff
@@ -0,0 +1,51 @@
+## Creating a new project
+1. Create a new folder that will house your `.csproj` and other project-related files.
+2. (EXTREMELY IMPORTANT) Inside this new folder, make a new folder for the source files of your project. For regular functionality-adding project this will be called `src`. However, if you are adding a different kind of project, it will be called something more applicable (ex. `test/` for a test project).
+3. In VS, Add a new `Solution Folder` in the new folder from the first step.
+4. Create the project via the VS `Add` menu (select the folder -> right click -> Add -> follow the wizard).
```
```suggestion
3. Open the `.slnf` you want to add your project to in VS. Then add a new `Solution Folder` in the new folder with the same name and location as the actual folder created in the first step.
4. Create the project via the VS `Add` menu (select the folder -> right click -> Add -> New Project... -> follow the wizard).
```

Daniel-Genkin-MS-2 committed:
Addressed more PR comments

Daniel-Genkin-MS-2 commented:
As basically everyone has reviewed this PR and I addressed all the comments, I am going to turn on auto-merge. Thanks everyone! This doc looks great! If anyone has any comments before it lands, please post them and pause auto-merge. I'll implement them as soon as I can. Otherwise, maybe we can do a follow up PR.

adityamandaleeka commented:
/azp run
